lib/internal/modules/esm/shared_constants.js

```diff
@@ -3,6 +3,40 @@
 
 'use strict';
 
+/*
+The shared memory area is divided in 4 32-bit long sections. It has to be 32-bit
+long as `Atomics.notify` only works with `Int32Array` objects.
+
+--32-bits-- --32-bits-- --32-bits-- --32-bits--
+    ^           ^           ^           ^
+    |           |           |           |
+    |           |           | NUMBER_OF_INCOMING_ASYNC_RESPONSES
+    |           |           |
+    |           | NUMBER_OF_INCOMING_MESSAGES
+    |           |
+    |   BOREDOM_OF_THREADS
+    |
+WORKER_TO_MAIN_THREAD_NOTIFICATION
+
+WORKER_TO_MAIN_THREAD_NOTIFICATION is only used to send notifications, its value is never used.
+
+BOREDOM_OF_THREADS does not represent a number, but an array of booleans.
+Currently, only two flags are used: WORKER_THREAD_IS_BORED, and MAIN_THREAD_IS_BORED
+
+xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx00: WORKER_THREAD_IS_BORED is false, and MAIN_THREAD_IS_BORED is false.
+xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx01: WORKER_THREAD_IS_BORED is true, and MAIN_THREAD_IS_BORED is false.
+xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx10: WORKER_THREAD_IS_BORED is false, and MAIN_THREAD_IS_BORED is true.
+xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx11: WORKER_THREAD_IS_BORED is true, and MAIN_THREAD_IS_BORED is true.
+
+NUMBER_OF_INCOMING_MESSAGES is used to communicate the number of messages sent
+by the main thread and not yet handled by the worker thread. It is also used to
+wake up the worker thread when it locks itself.
+
+NUMBER_OF_INCOMING_ASYNC_RESPONSES is used to let the main thread keep count of
+how many async responses to expect.
+
+*/
+
 module.exports = {
   WORKER_TO_MAIN_THREAD_NOTIFICATION: 0,
```

Comment:
I’m sure @joyeecheung would want these constants included in one of our existing snapshotted files rather than having us create yet another new file in this folder, for bootstrap performance.

Reply:
I'm not sure she would ;) `internal/modules/esm/shared_constants` is only used by `internal/modules/esm/hooks` and `internal/modules/esm/worker`, none of which is loaded at startup. `internal/modules/esm/hooks` is loaded conditionally by the `initializeHooks` function in `internal/modules/esm/utils` since we decided that the off-threading should happen only if there are user-provided hooks, so it wouldn't make sense inside the snapshot.
For readability and maintenance ease, I firmly believe it's best to leave those definitions in their own file.